### ISSUE TITLE
How can I use sass-embedded with tilt?

### ISSUE THREAD

**Issue by scivola** [ntkme/sass-embedded-host-ruby#28]
Before version 1.2.0, we can use sass-embedded with Tilt by using some workaround such as follows:

```rb
module Tilt
  class SassTemplate < Template
    self.default_mime_type = "text/css"

    @@load_paths = []

    def self.load_paths
      @@load_paths
    end

    begin
      require "sass-embedded"
    rescue LoadError => err
      raise err
    end

    def prepare
      @render_result = Sass.compile_string(data, **sass_options)
    rescue => e
      puts "ERROR"
      p e.sass_message, e.sass_stack, e.span
    end

    def evaluate(scope, locals, &block)
      @output ||= @render_result.css
    end

    def allows_script?
      false
    end

    private

    def sass_options
      # options.merge(filename: eval_file, line: line, syntax: :indented)
      options.merge(load_paths: @@load_paths, syntax: :indented)#load_paths: [])
    end
  end

  class ScssTemplate < SassTemplate
    self.default_mime_type = "text/css"

    def sass_options
      options.merge(syntax: :scss)#load_paths: [])
    end
  end
end
```

The above code overwrites the original SassTemplate and ScssTemplate of Tilt.

However, the following exception occurs at [here](https://github.com/rtomayko/tilt/blob/master/lib/tilt/sass.rb#L11-L21) after version 1.2.0.

```
LoadError: cannot load such file -- sassc
```

How can I use sass-embedded with Tilt?

**Comment by ntkme** [comment]
Due to the naming conflict with original "sass" gem, this gem no longer support `require 'sass'`, instead, it must be `require 'sass-embedded'`.

Instead of hacking tilt, please try this: https://github.com/ntkme/sassc-embedded-polyfill-ruby

**Comment by scivola** [comment]
Thank you for your quick advice.

I put 

```rb
gem 'sassc', github: 'sass/sassc-ruby', ref: 'refs/pull/233/head'
gem "sassc-embedded"
```

in my Gemfile, instead of 

```rb
gem "sass-embedded"
```

This causes:

```
/Users/XXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/async.rb:56:in `block in await': Response ID 0 doesn't match any outstanding requests. (Sass::CompileError)
```

Sorry, my code is a little complicated, so I cannot immediately show you a minimal reproduction code.

**Comment by ntkme** [comment]
To use the polyfill, you need to remove your own SassTemplate and just use what Tilt’s default SassC support.

**Comment by scivola** [comment]
Of course I removed my SassTemplate (I forgot to write it).

I can show the following part of the entire back trace:

```
/Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/async.rb:56:in `block in await': Response ID 0 doesn't match any outstanding requests. (Sass::CompileError)
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/async.rb:53:in `synchronize'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/async.rb:53:in `await'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/host.rb:135:in `block in await'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/host.rb:129:in `synchronize'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/host.rb:129:in `await'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded/host.rb:43:in `compile_request'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded.rb:199:in `compile_string'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sass-embedded-1.2.3/lib/sass/embedded.rb:45:in `compile_string'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/sassc-embedded-1.1.1/lib/sassc/embedded.rb:18:in `render'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/tilt-2.0.10/lib/tilt/sass.rb:28:in `evaluate'
        from /Users/XXXXX/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/tilt-2.0.10/lib/tilt/template.rb:109:in `render'
```

The exception seems occur only when two Sass files are processed simultaneously.
The template rendering is caused by a Rack request. So if an HTML page includes two links to CSSs that are generated by sass-embedded, simultanaous Sass rendering occurs.

**Comment by ntkme** [comment]
Looks like a race condition bug on the dart-sass-embedded rather on the host side: https://github.com/sass/dart-sass-embedded/blob/31d0eb695e01ba061deeb8767196d3d967f9569c/lib/src/dispatcher.dart#L161-L171

My guess is that the function referenced above got parallel execution and returned id 0 twice.

**Comment by ntkme** [comment]
@scivola Can you please try modify the following line in you local installation and see if the error goes away. This would help confirming the root cause.

``` diff
diff --git a/lib/sass/embedded/dispatcher.rb b/lib/sass/embedded/dispatcher.rb
index 5f2600f..b735021 100644
--- a/lib/sass/embedded/dispatcher.rb
+++ b/lib/sass/embedded/dispatcher.rb
@@ -11,7 +11,7 @@ module Sass
       def initialize
         @compiler = Compiler.new
         @observers = {}
-        @id = 0
+        @id = 0xfffffffe
         @mutex = Mutex.new
 
         Thread.new do
```

**Comment by ntkme** [comment]
I was able to reproduce this locally. This is indeed a bug in dart-sass-embedded compiler, meanwhile please give the patch mentioned in the previous comment a try, it should workaround this issue temporarily with the performance penalty of spawning and closing compiler for each request.

``` ruby
require 'sass-embedded'

10.times do
  Thread.new do
    Sass.compile_string('@import "not-found.scss"', importers: [
      {
        find_file_url: ->(*) { }
      }
    ])
  end
end

Thread.list.each do |t|
  t.join unless t == Thread.current
end
```

**Comment by ntkme** [comment]
https://github.com/sass/dart-sass-embedded/issues/79

**Comment by scivola** [comment]
I tried the above patch.
It solved the problem!

**Comment by ntkme** [comment]
Upstream PR to fix this issue: https://github.com/sass/dart-sass-embedded/pull/80

**Comment by ntkme** [comment]
Released v1.2.4 with the workaround mentioned earlier. Note that this workaround does have a performance impact.

I will release another version with the workaround as soon as the upstream bug fix lands.
